syncmaster/worker/handlers/file/base.py

```diff
@@ -59,14 +60,74 @@ def read(self) -> DataFrame:
         return df
 
     def write(self, df: DataFrame) -> None:
+        tmp_path = os.path.join(self.transfer_dto.directory_path, ".tmp", str(self.run_dto.id))
         writer = FileDFWriter(
-            connection=self.connection,
+            connection=self.df_connection,
             format=self.transfer_dto.file_format,
-            target_path=self.transfer_dto.directory_path,
+            target_path=tmp_path,
             options=self.transfer_dto.options,
         )
+        writer.run(df=df)
+
+        self._rename_files(tmp_path)
+
+        mover = FileMover(
+            connection=self.connection,
+            source_path=tmp_path,
+            target_path=self.transfer_dto.directory_path,
+        )
+        mover.run()
+
+    def _rename_files(self, tmp_path: str) -> None:
+        files = self.connection.list_dir(tmp_path)
+
+        for index, file_name in enumerate(files):
+            extension = self._get_file_extension(str(file_name))
+            new_name = self._get_file_name(str(index), extension)
+            old_path = os.path.join(tmp_path, file_name)
+            new_path = os.path.join(tmp_path, new_name)
+            self.connection.rename_file(old_path, new_path)
+
+    def _get_file_name(self, index: str, extension: str) -> str:
+        return (
+            self.transfer_dto.file_name_template.replace(
+                "{index}",
+                index,
+            )
+            .replace(
+                "{extension}",
+                extension,
+            )
+            .replace(
+                "{run_id}",
+                str(self.run_dto.id),
+            )
+            .replace(
+                "{run_created_at}",
+                self.run_dto.created_at.strftime("%Y_%m_%d_%H_%M_%S"),
```

Comment:
BTW, if ":" symbol is not allowed by Spark in file paths, shouldn't we add validation for `file_name_template` API schema?